TSBPD, TLPKTDROP, Drift Management (#21)
Haivision/srt-rfc: draft-sharabayko-mops-srt.md

during which a number of activities take place, such as report of ACKs({{packet-acks}})
or NAKs({{packet-naks}}).
Latency is configured through capability exchange during extended handshake process 
between initiator and responder. The handshake extension ({{handshake-extension-msg}}) has 
receiver and sender TSBPD delay information in milliseconds ({{tsbpd}}). The maximum value of latencies
from initiator and responder will be established. 

## Timestamp Based Packet Delivery {#tsbpd}

The initial intent of the SRT Timestamp Based Packet Delivery (TSBPD) feature is
to reproduce the output of the sending application at the input of the receiving one

> in live data transmission  mode (see {{data-transmission-mode}}).

Reviewer: ```suggestion
The goal of the SRT Timestamp Based Packet Delivery (TSBPD) mechanism is
to reproduce the input timing of the sending side in the output timing of the receiving side
in live data transmission mode (see {{data-transmission-mode}}).
```

It is designed in a way that SRT receiver, using the timestamp of the SRT data packet header,

> delivers packets to a receiving application at the same pace they were provided to the SRT sender by a sending applicaton.

Reviewer: ```suggestion
The SRT receiver, using the timestamp of the SRT data packet header,
delivers packets to a receiving application with a fixed minimum delay from the time the data packet was scheduled for sending on the sender's side.
```

(compensating for the time drift or different time zone) before releasing the packet to the application.
Packets can be withheld by the SRT receiver for a configured receiver delay.
Higher delay can accommodate a larger uniform packet drop rate or larger packet burst drop.
Packets received after their "play time" are dropped if Too-Late Packet Drop feature is enabled (see {{too-late-packet-drop}}).

The packet timestamp (in microseconds) is relative to the SRT connection creation time.
It is worth mentioning that the use of the packet sending time to stamp the packets is
inappropriate for TSBPD feature since a new time (current sending time) is used for retransmitted packets,
putting them out of order when inserted at their proper place in the stream. Packets are
inserted based on the sequence number in the header field. The origin time (in microseconds)
of the packet is already sampled when a packet is first submitted by the application to the SRT sender.
The TSBPD feature uses this time to stamp the packet for first transmission and any subsequent retransmission.
This timestamp and the configured SRT latency control the recovery buffer size and the instant that packets

> are delivered at the destination.

Reviewer: ```suggestion
The origin time (in microseconds)
of the packet is already sampled when a packet is first submitted by the application to the SRT sender.
The TSBPD mechanism uses this time to stamp the packet for first transmission and any subsequent retransmission.
This timestamp and the configured SRT latency control the recovery buffer size and the instant that packets
are delivered at the destination.
```

                |           |                   |                         |
                |           |                   |                         |
                |           |                   |                         |
      _____ Ready to       Sent              Received                 Ready to be
     /      be sent         |                   |                     delivered
  Packet        |           |                   |                         |
  State         |           |                   |                         |
                |           |                   |                         |
                |           |                   |                         |
                ----------------------------------------------------------------->
                                                                                Time
~~~
{: #fig-latency-points title="Key Latency Points during the Packet Transmission"}

The main packet states shown at in {{fig-latency-points}} are the following:

- "Ready to be sent": the packet is commited by the sending application, stamped and ready to be sent;

> - "Sent": the packet is passed to the UDP socket and sent;

Reviewer: ```suggestion
- "Scheduled for sending": the packet is committed by the sending application, stamped and ready to be sent;
- "Sent": the packet is passed to the UDP socket and sent;
```

- "Received": the packet is received and read from the UDP socket;
- "Ready to be delivered": the packet is ready to be delivered to the upstream (receiving) application,
  delivered and read by the application (see {{packet-delivery-time}}).

It is worth noting that the round-trip time (RTT) of an SRT link may
vary in time. However the actual end-to-end latency on the link becomes

> fixed and approximately equal to (RTT_0/2 + SRT Latency) once the SRT handshake exchange happens,

Reviewer: ```suggestion
fixed and is approximately equal to (RTT_0/2 + SRT Latency) once the SRT handshake exchange happens,
```

where RTT_0 is the actual value of the round-trip time during the SRT handshake
exchange (the value of the round-trip time once the SRT connection has been established).

The value of the sending delay is relatively small and evaluated in microseconds,

> in contrast to RTT/2 and SRT latency which are usually measured in milliseconds.

Reviewer: ```suggestion
The value of the sending delay depends on hardware performance. Usually, it is small (several microseconds) in comparison with RTT_0/2 and SRT latency usually measured in milliseconds.
```

on a number of packets rather than time duration to ensure enough
samples, independently of the media stream packet rate. The effect of
network jitter on the estimated time drift is attenuated by using a
large number of samples. The actual time drift being very slow (affecting a
stream only after many hours) does not require a fast reaction.

The receiver uses local time to be able to schedule events — to
determine, for example, if it's time to deliver a certain packet
right away. The timestamps in the packets themselves are just
references to the beginning of the session. When a packet is received
(with a timestamp from the sender), the receiver makes a reference to
the beginning of the session to recalculate its timestamp. The start
time is derived from the local time at the moment that the session is
connected. A packet timestamp equals "now" minus "StartTime", where
zthe latter is the point in time when the socket was created.

## Too-Late Packet Drop {#too-late-packet-drop}

Too-Late Packet Drop (TLPKTDROP) mechanism allows the sender to drop
packets that have no chance to be delivered in time and the receiver

> to skip missing packets that have not been delivered in time.

Reviewer: The timeout to drop a packet is based on the TSBPD mechanism.
